pydra/utils/hash.py

```diff
@@ -0,0 +1,330 @@
+    yield val
+
+
+@register_serializer
+def bytes_repr_float(obj: float, cache: Cache) -> Iterator[bytes]:
+    yield b"float:"
+    yield struct.pack("<d", obj)
+
+
+@register_serializer
+def bytes_repr_complex(obj: complex, cache: Cache) -> Iterator[bytes]:
+    yield b"complex:"
+    yield struct.pack("<dd", obj.real, obj.imag)
+
+
+@register_serializer
+def bytes_repr_dict(obj: dict, cache: Cache) -> Iterator[bytes]:
+    yield b"dict:{"
+    yield from bytes_repr_mapping_contents(obj, cache)
+    yield b"}"
+
+
+@register_serializer(_SpecialForm)
+@register_serializer(type)
+def bytes_repr_type(obj: type, cache: Cache) -> Iterator[bytes]:
+    cls = type(obj)
+    yield f"{cls.__module__}.{cls.__name__}".encode()
+
+
+@register_serializer(list)
+@register_serializer(tuple)
+def bytes_repr_seq(obj: Sequence, cache: Cache) -> Iterator[bytes]:
+    yield f"{obj.__class__.__name__}:(".encode()
+    yield from bytes_repr_sequence_contents(obj, cache)
+    yield b")"
+
+
+@register_serializer(set)
+@register_serializer(frozenset)
+def bytes_repr_set(obj: Set, cache: Cache) -> Iterator[bytes]:
+    yield f"{obj.__class__.__name__}:{{".encode()
+    yield from bytes_repr_sequence_contents(sorted(obj), cache)
+    yield b"}"
+
+
+def bytes_repr_mapping_contents(mapping: Mapping, cache: Cache) -> Iterator[bytes]:
+    """Serialize the contents of a mapping
+
+    Concatenates byte-serialized keys and hashed values.
+
+    .. code-block:: python
+
+        >>> from pydra.utils.hash import bytes_repr_mapping_contents, Cache
+        >>> generator = bytes_repr_mapping_contents({"a": 1, "b": 2}, Cache({}))
+        >>> b''.join(generator)
+        b'str:1:a=...str:1:b=...'
+    """
+    for key in sorted(mapping):
+        yield from bytes_repr(key, cache)
+        yield b"="
+        yield bytes(hash_single(mapping[key], cache))
+
+
+def bytes_repr_sequence_contents(seq: Sequence, cache: Cache) -> Iterator[bytes]:
+    """Serialize the contents of a sequence
+
+    Concatenates hashed values.
+
+    .. code-block:: python
+
+        >>> from pydra.utils.hash import bytes_repr_sequence_contents, Cache
+        >>> generator = bytes_repr_sequence_contents([1, 2], Cache({}))
+        >>> list(generator)
+        [b'\x6d...', b'\xa3...']
+    """
+    for val in seq:
+        yield bytes(hash_single(val, cache))
+
+
+if HAVE_NUMPY:
+
+    @register_serializer(numpy.generic)
+    @register_serializer(numpy.ndarray)
+    def bytes_repr_numpy(obj: numpy.ndarray, cache: Cache) -> Iterator[bytes]:
+        yield f"{obj.__class__.__module__}{obj.__class__.__name__}:{obj.size}:".encode()
+        if obj.dtype == "object":
+            yield from bytes_repr_sequence_contents(iter(obj.ravel()), cache)
+        else:
+            bytes_it = iter(obj.tobytes(order="C"))
+            for chunk in iter(
+                lambda: bytes(itertools.islice(bytes_it, NUMPY_CHUNK_LEN)), b""
+            ):
+                yield chunk
```

Comment:
We're already allocating the RAM, unless there's a speedup I don't understand, I think we should keep this simple.

```suggestion
            yield obj.tobytes(order="C")
```

Reply:
Oh, I didn't know that RAM allocation was the rate-limiting factor when trying to hash large objects. I assumed it was algorithmic but it is a bit of a black box to me